File: Quickunion.cpp
#include <vector>
class Quickunion {
    private:
        std::vector<int> id;
        std:: vector<int> size;


    public:
        Quickunion(int n){
            id.resize(n); // resize the array
            for(int i=0;i<n;i++){
                id[i]=0;

            }
        }
        int find(int p){
            while(id[p] != p){
                p =id[p];
            }
            return p;
        }
        void QuickUnion(int p, int q){
           int proot = find(p);
           int rroot = find(q);
           id[proot]=rroot;
        }

        bool connected(int p, int q){
            return find(p) == find(q);
        }

        int pathcompressedfind(int p){
            while(id[p] != p){
                id[p]= id[id[p]];
                p= id[p];
            }
            return p;
        }

        void weightedcompressedfile(int p , int q)
        {
            int pr = find(p);
            int qr = find (q);
            if(pr ==qr)return;
            if(size[pr]<size[qr]){id[pr]=qr;size[qr] += size[pr];}
            else{id[qr]=pr;size[pr] += size[qr];}
        }
};
int main(){
    Quickunion(20);

    return 0;
}

File: Quickunion.py
class Quickunionuf:
    def __init__(self,n):
        #initialisiation of the array
        self.id = [i for i in range(n)]
        self.size = [1]* n
    
    def find(self,p):
        while(self.id[p] != p):
            p = self.id[p]
        return p
    
    def union(self,p,q):
        proot= self.find(p)
        rroot= self.find(q)
        self.id[proot]=rroot
    
    def connected(self,p,q):
        return self.find(p)==self.find(q)
    
    def weightedunion(self,p,q):
        proot = self.find(p)
        rroot = self.find(q)
        if proot==rroot: return
        if self.size[proot]< self.size[rroot]:
            self.id[proot]=rroot
            self.size[rroot] += self.size[proot]
        else:
             self.id[rroot]=proot
             self.size[proot] += self.size[rroot]
    
    def pathcompressedfind(self,p):
        while(self.id[p]!=p):
            self.id[p] =self.id[self.id[p]]
            p=self.id[p]
        return p




